misc-utils/lsfd-sock-xinfo.c

```diff
@@ -726,3 +806,124 @@ static void load_xinfo_from_proc_udp(ino_t netns_inode)
 				     "/proc/net/udp",
 				     &udp_xinfo_class);
 }
+
+/*
+ * RAW
+ */
+struct raw_xinfo {
+	struct l3_xinfo l3;
+	uint16_t protocol;
+};
+
+static char *raw_get_name(struct sock_xinfo *sock_xinfo,
+			  struct sock *sock  __attribute__((__unused__)))
+{
+	struct l3_xinfo_class *class = (struct l3_xinfo_class *)sock_xinfo->class;
+	char *str = NULL;
+	unsigned int st = ((struct l3_xinfo *)sock_xinfo)->st;
+	void *laddr = class->get_laddr((struct l3_xinfo *)sock_xinfo);
+	void *raddr = class->get_raddr((struct l3_xinfo *)sock_xinfo);
+	struct raw_xinfo *raw = ((struct raw_xinfo *)sock_xinfo);
+	size_t ssize = class->size_as_cstring;
+	char *local_s = alloca(ssize);
+	char *remote_s = alloca(ssize);
+	const char *st_str = tcp_decode_state(st);
+
+	if (!inet_ntop(class->family, laddr, local_s, ssize))
+		xasprintf(&str, "state=%s", st_str);
+	else if (class->is_any_addr(raddr)
+		 || !inet_ntop(class->family, raddr, remote_s, ssize))
+		xasprintf(&str, "state=%s protocol=%u laddr=%s",
+			  st_str,
+			  raw->protocol, local_s);
+	else
+		xasprintf(&str, "state=%s protocol=%u laddr=%s raddr=%s",
+			  st_str,
+			  raw->protocol, local_s, remote_s);
+	return str;
+}
+
+static char *raw_get_type(struct sock_xinfo *sock_xinfo __attribute__((__unused__)),
+			  struct sock *sock __attribute__((__unused__)))
+{
+	return strdup("raw");
+}
+
+static bool raw_fill_column(struct proc *proc __attribute__((__unused__)),
+			    struct sock_xinfo *sock_xinfo,
+			    struct sock *sock __attribute__((__unused__)),
+			    struct libscols_line *ln __attribute__((__unused__)),
+			    int column_id,
+			    size_t column_index __attribute__((__unused__)),
+			    char **str)
+{
+	l2_fill_column_handler(INET, sock_xinfo, column_id, str);
+
+	if (column_id == COL_RAW_PROTOCOL)
+		xasprintf(str, "%u",
+			  (unsigned int)(((struct raw_xinfo *)sock_xinfo)->protocol));
+	else
+		return false;
+
+	return true;
+}
+
+static struct sock_xinfo *raw_xinfo_scan_line(const struct l3_xinfo_class *class,
+					      char * line,
+					      ino_t netns_inode,
+					      enum sysfs_byteorder byteorder)
+{
+	unsigned long local_addr;
+	unsigned long protocol;
+	unsigned long remote_addr;
+	unsigned long st;
+	unsigned long long inode;
+	struct raw_xinfo *raw;
+	struct inet_xinfo *inet;
+	struct sock_xinfo *sock;
+
+	if (sscanf(line, "%*d: %lx:%lx %lx:%*x %lx %*x:%*x %*x:%*x %*x %*u %*u %lld",
+		   &local_addr, &protocol, &remote_addr,
+		   &st, &inode) != 5)
+		return NULL;
+
+	if (inode == 0)
+		return NULL;
+
+	raw = xcalloc(1, sizeof(struct raw_xinfo));
+	inet = (struct inet_xinfo *)raw;
+	sock = (struct sock_xinfo *)inet;
+	sock->class = (struct sock_xinfo_class *)class;
+	sock->inode = (ino_t)inode;
+	sock->netns_inode = netns_inode;
+	inet->local_addr.s_addr = kernel32_to_cpu(byteorder, local_addr);
+	inet->remote_addr.s_addr = kernel32_to_cpu(byteorder, remote_addr);
+	raw->protocol = protocol;
+	((struct l3_xinfo *)raw)->st = st;
```

Comment:
This could work with a lot less unchecked casts:
```diff
diff --git a/misc-utils/lsfd-sock-xinfo.c b/misc-utils/lsfd-sock-xinfo.c
index 61f9c0ba653a..f911693b8f71 100644
--- a/misc-utils/lsfd-sock-xinfo.c
+++ b/misc-utils/lsfd-sock-xinfo.c
@@ -891,15 +891,15 @@ static struct sock_xinfo *raw_xinfo_scan_line(const struct l3_xinfo_class *class
                return NULL;
 
        raw = xcalloc(1, sizeof(struct raw_xinfo));
-       inet = (struct inet_xinfo *)raw;
-       sock = (struct sock_xinfo *)inet;
+       inet = &raw->l3.inet;
+       sock = &inet->sock;
        sock->class = (struct sock_xinfo_class *)class;
        sock->inode = (ino_t)inode;
        sock->netns_inode = netns_inode;
        inet->local_addr.s_addr = kernel32_to_cpu(byteorder, local_addr);
        inet->remote_addr.s_addr = kernel32_to_cpu(byteorder, remote_addr);
        raw->protocol = protocol;
-       ((struct l3_xinfo *)raw)->st = st;
+       raw->l3.st = st;
 
        return sock;
 }
```

Reply:
Done.